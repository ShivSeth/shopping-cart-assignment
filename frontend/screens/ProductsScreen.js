import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ProductCard from '../components/ProductCard';
import { fetchProducts } from '../actions/productsAction';
import { fetchCategory } from '../actions/categoryActions';
import './ProductsScreen.scss';
import { useHistory, useLocation } from 'react-router';
import Dropdown from '../components/Dropdown';

const ProductsScreen = () => {
  const history = useHistory();
  const location = useLocation();

  const products = useSelector((state) => state.products);
  const category = useSelector((state) => state.category);
  const dispatch = useDispatch();

  const [selected, setSelected] = useState({ name: 'Select Category' });

  let sectionId = location.search.split('=')[1];
  let { categoryList } = category;
  let { error, loading, productsList } = products;

  useEffect(() => {
    dispatch(fetchCategory());
    dispatch(fetchProducts());

    if (selected.id) {
      history.push(`?section=${selected.id}`);
    }
  }, [dispatch, selected]);
  return (
    <div className="products-page">
      <div className="side-menu">
        <ul className="menu-options">
          {categoryList ? (
            categoryList
              .filter((category) => category.enabled)
              .map((category) => (
                <li
                  tabIndex={0}
                  className={category.id === sectionId ? 'active-section' : ''}
                  key={category.id}
                  onClick={() => {
                    if (sectionId && sectionId === category.id) {
                      history.push('/products');
                    } else {
                      history.push(`?section=${category.id}`);
                    }
                  }}
                >
                  {category.name}
                </li>
              ))
          ) : (
            <p>Loading...</p>
          )}
        </ul>
      </div>
      <div className="side-menu-dropdown">
        {categoryList ? (
          <Dropdown
            selected={selected.name}
            setSelected={setSelected}
            dropdownOptions={categoryList.filter((c) => c.enabled)}
          />
        ) : (
          <p>Loading...</p>
        )}
      </div>
      <div className="product-content">
        {productsList ? (
          <div className="products-list">
            {productsList
              .filter((product) => {
                return sectionId ? sectionId === product.category : product;
              })
              .map((product) => (
                <ProductCard key={product.id} product={product} />
              ))}
          </div>
        ) : error ? (
          <p>Error</p>
        ) : (
          <p>loading</p>
        )}
      </div>
    </div>
  );
};

export default ProductsScreen;
