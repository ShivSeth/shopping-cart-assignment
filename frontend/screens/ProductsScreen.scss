.products-page {
  margin: 0 auto;
  width: 90%;
  display: flex;
  .side-menu-dropdown {
    display: none;
    width: 100%;
  }

  .side-menu {
    background-color: lightgray;
    flex-basis: 15%;
    min-width: 175px;
    align-self: stretch;
    .menu-options {
      .active-section {
        background-color: darkgray;
      }

      li {
        padding: 0.75rem;
        border-bottom: 1px solid gray;
        &:hover {
          cursor: pointer;
          background-color: rgb(141, 135, 135);
        }
      }
    }
  }
  .product-content {
    flex-basis: 85%;

    display: flex;
    justify-content: center;
  }
  .products-list {
    background-color: white;
    display: flex;
    flex-wrap: wrap;
    margin-bottom: 20;
  }
}

@media screen and (max-width: 675px) {
  .products-page {
    width: 100%;
    display: block;

    .side-menu {
      display: none;
    }

    .side-menu-dropdown {
      display: block;
    }

    .product-content {
      display: block;
    }

    .products-list {
      justify-content: center;
    }
  }
}
