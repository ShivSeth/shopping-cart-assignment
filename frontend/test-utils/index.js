import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { AuthContextProvider } from '../context/AuthContext';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import store from '../store';
store;
const AllTheProviders = ({ children }) => {
  const history = createMemoryHistory();

  return (
    <Provider store={store}>
      <AuthContextProvider>
        <Router history={history}>{children}</Router>
      </AuthContextProvider>
    </Provider>
  );
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
