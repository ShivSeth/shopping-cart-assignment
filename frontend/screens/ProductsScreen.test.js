import { render, screen } from '../test-utils';
import ProductsScreen from './ProductsScreen.js';
import userEvent from '@testing-library/user-event';

describe('<ProductsScreen/>', () => {
  test('should Products Add to cart button be present', async () => {
    const { container } = render(<ProductsScreen />);

    const AddToCartButtons = await screen.findAllByRole('button', {
      name: 'Add to cart',
    });
    expect(AddToCartButtons.length).toBe(3);
    expect(container).toMatchSnapshot();
  });
  test('When a product is added product screen display it ', async () => {
    const { container } = render(<ProductsScreen />);
    const addToCartButtons = await screen.findAllByRole('button', {
      name: 'Add to cart',
    });
    userEvent.click(addToCartButtons[0]);
    expect(screen.getAllByText(/1 in cart/).length).toBe(2);
    expect(container).toMatchSnapshot();
  });
});
