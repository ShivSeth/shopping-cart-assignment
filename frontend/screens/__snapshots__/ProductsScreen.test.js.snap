// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductsScreen/> When a product is added product screen display it  1`] = `
<div>
  <div
    class="products-page"
  >
    <div
      class="side-menu"
    >
      <ul
        class="menu-options"
      >
        <li
          class=""
          tabindex="0"
        >
          Beverages
        </li>
        <li
          class=""
          tabindex="0"
        >
          Bakery Cakes and Dairy
        </li>
        <li
          class=""
          tabindex="0"
        >
          Beauty and Hygiene
        </li>
      </ul>
    </div>
    <div
      class="side-menu-dropdown"
    >
      <div
        class="dropdown"
      >
        <div
          class="dropdown-btn"
        >
          Select Category
           
          <span
            class="fas fa-caret-down"
          />
        </div>
      </div>
    </div>
    <div
      class="product-content"
    >
      <div
        class="products-list"
      >
        <div
          class="card-holder"
        >
          <h4>
            Tata Tea Gold Leaf Tea, 500 gm
          </h4>
          <div
            class="item-details"
          >
            <div
              class="image-description"
            >
              <img
                alt="Tata Tea Gold Leaf Tea, 500 gm"
                src="/static/images/products/beverages/tata-tea.jpg"
              />
              <p>
                Tata Tea Gold is a unique blend of fine Assam tea leaves with special 15%  long leaf.
              </p>
            </div>
            <div
              class="price-options"
            >
              <h5>
                MRP Rs 165
              </h5>
              <button>
                1 in cart
              </button>
            </div>
            <div
              class="price-options-button"
            >
              <button>
                1 in cart
                 @Rs 165
              </button>
            </div>
          </div>
        </div>
        <div
          class="card-holder"
        >
          <h4>
            Sandwich Bread - White, Chemical Free, 400 gm
          </h4>
          <div
            class="item-details"
          >
            <div
              class="image-description"
            >
              <img
                alt="Sandwich Bread - White, Chemical Free, 400 gm"
                src="/static/images/products/bakery-cakes-dairy/bread.jpg"
              />
              <p>
                Freshly Baked bread is one of lifes greatest simple pleasures and at Bigbasket we decided to give you just that. We start baking our breads once you order using the finest ingredients available.
              </p>
            </div>
            <div
              class="price-options"
            >
              <h5>
                MRP Rs 32
              </h5>
              <button>
                Add to cart
              </button>
            </div>
            <div
              class="price-options-button"
            >
              <button>
                Add to cart
                 @Rs 32
              </button>
            </div>
          </div>
        </div>
        <div
          class="card-holder"
        >
          <h4>
            Listerine Mouthwash - Cool Mint, 250 ml
          </h4>
          <div
            class="item-details"
          >
            <div
              class="image-description"
            >
              <img
                alt="Listerine Mouthwash - Cool Mint, 250 ml"
                src="/static/images/products/beauty-hygiene/listerine.jpg"
              />
              <p>
                Listerine Cool mint Antibacterial Mouthwash is rapid & easy to use, a 30 second "slosh" with Listerine after brushing and flossing reaches parts of the mouth that may be missed. Used two times daily, Listerine provides 24-hour defence against plaque and lasting bright breath assurance.
              </p>
            </div>
            <div
              class="price-options"
            >
              <h5>
                MRP Rs 109
              </h5>
              <button>
                Add to cart
              </button>
            </div>
            <div
              class="price-options-button"
            >
              <button>
                Add to cart
                 @Rs 109
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ProductsScreen/> should Products Add to cart button be present 1`] = `
<div>
  <div
    class="products-page"
  >
    <div
      class="side-menu"
    >
      <ul
        class="menu-options"
      >
        <li
          class=""
          tabindex="0"
        >
          Beverages
        </li>
        <li
          class=""
          tabindex="0"
        >
          Bakery Cakes and Dairy
        </li>
        <li
          class=""
          tabindex="0"
        >
          Beauty and Hygiene
        </li>
      </ul>
    </div>
    <div
      class="side-menu-dropdown"
    >
      <div
        class="dropdown"
      >
        <div
          class="dropdown-btn"
        >
          Select Category
           
          <span
            class="fas fa-caret-down"
          />
        </div>
      </div>
    </div>
    <div
      class="product-content"
    >
      <div
        class="products-list"
      >
        <div
          class="card-holder"
        >
          <h4>
            Tata Tea Gold Leaf Tea, 500 gm
          </h4>
          <div
            class="item-details"
          >
            <div
              class="image-description"
            >
              <img
                alt="Tata Tea Gold Leaf Tea, 500 gm"
                src="/static/images/products/beverages/tata-tea.jpg"
              />
              <p>
                Tata Tea Gold is a unique blend of fine Assam tea leaves with special 15%  long leaf.
              </p>
            </div>
            <div
              class="price-options"
            >
              <h5>
                MRP Rs 165
              </h5>
              <button>
                Add to cart
              </button>
            </div>
            <div
              class="price-options-button"
            >
              <button>
                Add to cart
                 @Rs 165
              </button>
            </div>
          </div>
        </div>
        <div
          class="card-holder"
        >
          <h4>
            Sandwich Bread - White, Chemical Free, 400 gm
          </h4>
          <div
            class="item-details"
          >
            <div
              class="image-description"
            >
              <img
                alt="Sandwich Bread - White, Chemical Free, 400 gm"
                src="/static/images/products/bakery-cakes-dairy/bread.jpg"
              />
              <p>
                Freshly Baked bread is one of lifes greatest simple pleasures and at Bigbasket we decided to give you just that. We start baking our breads once you order using the finest ingredients available.
              </p>
            </div>
            <div
              class="price-options"
            >
              <h5>
                MRP Rs 32
              </h5>
              <button>
                Add to cart
              </button>
            </div>
            <div
              class="price-options-button"
            >
              <button>
                Add to cart
                 @Rs 32
              </button>
            </div>
          </div>
        </div>
        <div
          class="card-holder"
        >
          <h4>
            Listerine Mouthwash - Cool Mint, 250 ml
          </h4>
          <div
            class="item-details"
          >
            <div
              class="image-description"
            >
              <img
                alt="Listerine Mouthwash - Cool Mint, 250 ml"
                src="/static/images/products/beauty-hygiene/listerine.jpg"
              />
              <p>
                Listerine Cool mint Antibacterial Mouthwash is rapid & easy to use, a 30 second "slosh" with Listerine after brushing and flossing reaches parts of the mouth that may be missed. Used two times daily, Listerine provides 24-hour defence against plaque and lasting bright breath assurance.
              </p>
            </div>
            <div
              class="price-options"
            >
              <h5>
                MRP Rs 109
              </h5>
              <button>
                Add to cart
              </button>
            </div>
            <div
              class="price-options-button"
            >
              <button>
                Add to cart
                 @Rs 109
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
